compiler/noirc_frontend/src/ownership/last_uses.rs

```diff
@@ -0,0 +1,410 @@
+    }
+
+    fn track_variables_in_for(&mut self, for_expr: &ast::For) {
+        self.track_variables_in_expression(&for_expr.start_range);
+        self.track_variables_in_expression(&for_expr.end_range);
+        self.track_variables_in_loop(&for_expr.block);
+    }
+
+    fn track_variables_in_while(&mut self, while_expr: &ast::While) {
+        self.track_variables_in_expression(&while_expr.condition);
+        self.track_variables_in_loop(&while_expr.body);
+    }
+
+    fn track_variables_in_loop(&mut self, loop_body: &Expression) {
+        self.push_loop_scope();
+        self.track_variables_in_expression(loop_body);
+        self.pop_loop_scope();
+    }
+
+    fn track_variables_in_if(&mut self, if_expr: &ast::If) {
+        self.track_variables_in_expression(&if_expr.condition);
+
+        self.branch(0);
+        self.track_variables_in_expression(&if_expr.consequence);
+        self.unbranch();
+
+        if let Some(alt) = &if_expr.alternative {
+            self.branch(1);
+            self.track_variables_in_expression(alt);
+            self.unbranch();
+        }
+    }
+
+    fn track_variables_in_match(&mut self, match_expr: &ast::Match) {
+        for (i, case) in match_expr.cases.iter().enumerate() {
+            for argument in &case.arguments {
+                self.declare_variable(*argument);
+            }
+
+            self.branch(i);
+            self.track_variables_in_expression(&case.branch);
+            self.unbranch();
+        }
+
+        if let Some(default_case) = &match_expr.default_case {
+            self.branch(match_expr.cases.len());
+            self.track_variables_in_expression(default_case);
+            self.unbranch();
+        }
+    }
+
+    fn track_variables_in_tuple(&mut self, elems: &[Expression]) {
+        for elem in elems {
+            self.track_variables_in_expression(elem);
+        }
+    }
+
+    fn track_variables_in_call(&mut self, call: &ast::Call) {
+        self.track_variables_in_expression(&call.func);
+        for arg in &call.arguments {
+            self.track_variables_in_expression(arg);
+        }
+    }
+
+    fn track_variables_in_let(&mut self, let_expr: &ast::Let) {
+        self.track_variables_in_expression(&let_expr.expression);
+        self.declare_variable(let_expr.id);
+    }
+
+    fn track_variables_in_constrain(
+        &mut self,
+        boolean: &Expression,
+        msg: &Option<Box<(Expression, crate::hir_def::types::Type)>>,
+    ) {
+        self.track_variables_in_expression(boolean);
+
+        if let Some(msg) = msg {
+            self.track_variables_in_expression(&msg.0);
+        }
+    }
+
+    fn track_variables_in_assign(&mut self, assign: &ast::Assign) {
+        self.track_variables_in_lvalue(&assign.lvalue);
+        self.track_variables_in_expression(&assign.expression);
+    }
+
+    fn track_variables_in_lvalue(&mut self, lvalue: &ast::LValue) {
+        match lvalue {
+            ast::LValue::Ident(_) => (),
```

Comment:
Do we need to handle this ident? If not, why can it be omitted?

Reply:
Nope - I'll add a comment explaining why not. It's because identifiers are never moved into an lvalue position. That position always expects a mutable reference since if the value is moved there'd be no point assigning to it since it could not be used any more.